### PR TITLE
[flutter_tools] add support for dart defines to flutter test

### LINKED ISSUE #34505: There is no way to test code guarded by `kReleaseMode`
The only way to test this code is to run it on a device in release mode, which is only possible on a physical device.

We should provide a way for flutter_test to set the constant to true.  However, we have to do so in a way that doesn't remove its const-ness, and doesn't prevent treeshaking from working.

Example:

```dart
import 'package:flutter/foundation.dart';

String foo() {
  if (kReleaseMode) {
    return null;
  }
  return _calculateABunchOfStuffAndReturnALongString();
}
```

Is only testable in the non-release branch currently.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Allow providing arbitrary dart defines to `flutter test`. this will also make it possible to unit test canvaskit once the frontend server change is in.

Fixes https://github.com/flutter/flutter/issues/34505 , would allow tests to provide --dart-define=<release mode define>

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/test.dart [hunk 1]
```diff
@@ -184,6 +185,18 @@ class TestCommand extends FlutterCommand {
     final String excludeTags = stringArg('exclude-tags');
     final BuildInfo buildInfo = await getBuildInfo(forcedBuildMode: BuildMode.debug);
 
+    if (buildInfo.packageConfig['test_api'] == null) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/test.dart:188]: Now that the package config is eagerly parsed this check can be done sooner

File: packages/flutter_tools/lib/src/test/runner.dart [hunk 2]
```diff
@@ -90,9 +90,6 @@ class _FlutterTestRunnerImpl implements FlutterTestRunner {
   }) async {
     // Configure package:test to use the Flutter engine for child processes.
     final String shellPath = globals.artifacts.getArtifactPath(Artifact.flutterTester);
-    if (!globals.processManager.canRun(shellPath)) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/test/runner.dart:93]: This shouldn't be necessary with the updated (years ago) artifacts check

**Review by dnfield — COMMENTED** [review]
LGTM

**Comment by Nikzed** [comment]
what's the point of it if you still can't set kDebugMode as a variable like `debugDefaultTargetPlatformOverride`?

**Pull request merged**